src/core/bpf-lsm.c

```diff
@@ -0,0 +1,335 @@
+        if (r)
+                return log_error_errno(r, "Failed to load BPF object");
+
+        /* Close dummy map */
+        inner_map_fd = safe_close(inner_map_fd);
+
+        *ret_obj = TAKE_PTR(obj);
+
+        return 0;
+}
+
+static int mac_bpf_use(void) {
+        _cleanup_free_ char *lsm_list = NULL;
+        static int cached_use = -1;
+        int r;
+
+        if (cached_use < 0) {
+                cached_use = 0;
+
+                r = read_one_line_file("/sys/kernel/security/lsm", &lsm_list);
+                if (r < 0) {
+                       if (errno != ENOENT)
+                               log_debug_errno(r, "Failed to read /sys/kernel/security/lsm: %m");
+
+                       return 0;
+                }
+
+                const char *p = lsm_list;
+
+                for (;;) {
+                        _cleanup_free_ char *word = NULL;
+
+                        r = extract_first_word(&p, &word, ",", 0);
+                        if (r == 0)
+                                break;
+                        if (r == -ENOMEM)
+                                return log_oom();
+                        if (r < 0)
+                                return 0;
+
+                        if (strstr(word, "bpf")) {
+                                cached_use = 1;
+                                break;
+                        }
+                }
+        }
+
+        return cached_use;
+}
+
+int lsm_bpf_supported(void) {
+        _cleanup_(restrict_fs_bpf_freep) struct restrict_fs_bpf *obj = NULL;
+        static int supported = -1;
+        int r;
+
+        if (supported >= 0)
+                return supported;
+
+        if (dlopen_bpf() < 0) {
+                log_info_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Failed to open libbpf, LSM BPF is not supported");
+                return supported = 0;
+        }
+
+        r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0) {
+                log_info_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+                return supported = 0;
+        }
+
+        if (r == 0) {
+                log_info_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                         "Not running with unified cgroup hierarchy, LSM BPF is not supported");
+                return supported = 0;
+        }
+
+        r = mac_bpf_use();
+        if (r < 0) {
+                log_info_errno(r, "Can't determine whether the BPF LSM module is used: %m");
+                return supported = 0;
+        }
+
+        if (r == 0) {
+                log_info_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                         "BPF LSM hook not enabled in the kernel, LSM BPF not supported");
+                return supported = 0;
+        }
+
+        r = prepare_restrict_fs_bpf(&obj);
+        if (r < 0)
+                return log_info_errno(r, "Failed to load BPF object: %m");
+
+        return bpf_can_link_lsm_program(obj->progs.restrict_filesystems);
```

Comment:
is it intentional that these 2 calls above don't update the cache flag?

Reply:
Oops, I missed those. Thanks!